Autosync - repeat syncs & GUI

tomasMizera opened the pull request:
PR adds two parts of Autosync: GUI and retry of sync when it fails. 

**Retry**  
Retry is handled by `SynchronizationManager`, it has an internal array of `syncProcesses` that holds a sync proceses during the whole time of sync (sync start, failure, waiting for another sync, sync again, finish).

**GUI**, some screenshots:

<table>
  <tr>
    <td><img src="https://user-images.githubusercontent.com/22449698/160770620-872e4141-1ce1-49d1-84ad-8c195201cb5d.jpeg" width="300"/></td>
    <td><img src="https://user-images.githubusercontent.com/22449698/160770622-771f69d0-a154-4a47-a734-792d4bdfe001.jpeg"  width="300"/></td>
  </tr>
  <tr>
    <td><img src="https://user-images.githubusercontent.com/22449698/160770625-0b4c48f9-c09d-43bd-952d-f0cff7e7647d.jpeg" width="300"/></td>
    <td><img src="https://user-images.githubusercontent.com/22449698/160770629-bbbdd649-6bcd-496e-a73f-4882e0182ad2.jpeg"  width="300"/></td>
  </tr>
  <tr>
    <td><img src="https://user-images.githubusercontent.com/22449698/160770599-a964eaea-d8fb-44be-be56-83c90ed7c71a.jpeg" width="300"/></td>
    <td><img src="https://user-images.githubusercontent.com/22449698/160770614-894b8f6f-9df8-4ee2-9935-1b02312807e2.jpeg"  width="300"/></td>
  </tr>
 </table>

PeterPetrik commented on app/qml/Banner.qml:
```diff
@@ -77,12 +97,9 @@ Rectangle {
-      textItem.text: "<style>a:link { color: " + banner.linkColor
-            + "; text-decoration: underline; }</style>" +
-            qsTr("%1<br><a href='%2' style=\"color: %3;\">Learn more</a>").arg(banner.text).arg(banner.link).arg(InputStyle.learnMoreLinkColor)
+      textItem.text: banner.getText()
```
Question: Does binding works if `getText()` is function? so if you change `banner.text` for example, does it automatically update `textItem.text`?

tomasMizera replied:
Yes yes, Qt goes over `js` functions and finds bindings there. Those functions are then reevaluated when some properties used in the function change

PeterPetrik replied:
good to know , thanks

PeterPetrik commented on app/qml/InputStyle.qml:
```diff
@@ -30,6 +30,8 @@ QtObject {
     property color labelColor: "#999999"
     property color secondaryFontColor: "#B3B3B3"
 
+    property color informationColor: "#86A3C3"
```
what is used for? maybe in InputStyle it could be more based on usage (like (secondary)backgroundColor?)

PeterPetrik commented on app/qml/banners/AutoHideBanner.qml:
```diff
@@ -0,0 +1,31 @@
+  Timer {
+    id: hideTimer
+
+    interval: 3000
```
3000 == cannot it be Banner property (so it is possible to set from outside). also constant used on 2 places

PeterPetrik commented on app/qml/dialogs/SyncFailedDialog.qml:
```diff
@@ -0,0 +1,23 @@
+
+  standardButtons: StandardButton.Ok
+
+  // It is also possible to set "detailedText"
```
Probably should be part of the doc for this file? like on line 12 add doxygen-like doc?

PeterPetrik commented on app/synchronizationerror.cpp:
```diff
@@ -0,0 +1,64 @@
+
+#include "synchronizationerror.h"
+
+SynchronizationError::SynchronizationError()
```
= default

PeterPetrik commented on app/synchronizationerror.cpp:
```diff
@@ -0,0 +1,64 @@
+{
+  switch ( errorType )
+  {
+    case AnotherProcessIsRunning: return true;
+
+    case VersionMismatch: return true;
+
+    case ServerError: return true;
```
```
  case AnotherProcessIsRunning
   case VersionMismatch: 
   case ServerError: 
return true;
```

PeterPetrik commented on app/synchronizationerror.h:
```diff
@@ -0,0 +1,44 @@
+    };
+    Q_ENUMS( ErrorType );
+
+    //! Parser error from server based on error code and message and returns one
```
Parser > Parses?

PeterPetrik commented on app/synchronizationerror.h:
```diff
@@ -0,0 +1,44 @@
+      ProjectNotFound,
+      VersionMismatch,
+      ServerError,
+      Other
```
I would use `UnknownError` instead of `Other`.

PeterPetrik commented on app/synchronizationerror.h:
```diff
@@ -0,0 +1,44 @@
+    static ErrorType errorType( int errorCode, const QString &errorMessage );
+
+    //! Returns true if the errorType can disappear in next sync try
+    static bool isWorthOfRetry( ErrorType errorType );
```
Not sure if I like the name... but cannot find the better one... maybe `possiblyTemporal` or `isPernament`

PeterPetrik commented on app/synchronizationoptions.h:
```diff
@@ -0,0 +1,38 @@
+
+    enum Strategy
+    {
+      Basic = 0, //! try to sync once, finishes on first failure
```
SingleShot? NoRetries?

PeterPetrik reviewed (commented):
nice job

tomasMizera commented:
@PeterPetrik addressed your comments here [b1ecb10](https://github.com/lutraconsulting/input/pull/2004/commits/b1ecb10f9b53b08cd1f1bac4e2b560ec6a74c0fe)

Can we merge it to integration branch and leave it there for further testing?

PeterPetrik reviewed (approved):
👏
